noodles-bed/src/de.rs

```diff
@@ -0,0 +1,939 @@
+
+use noodles_core::Position;
+
+use serde::Deserialize;
+use serde::de::{
+    self, DeserializeSeed, EnumAccess, IntoDeserializer, MapAccess, SeqAccess,
+    VariantAccess, Visitor,
+};
+
+use crate::{Record, reader::Reader, error::{Error, Result}};
+
+impl<'de> Deserialize<'de> for Record<3> {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        enum Field { Chrom, Start, End }
+
+        impl<'de> Deserialize<'de> for Field {
+            fn deserialize<D>(deserializer: D) -> std::result::Result<Field, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                struct FieldVisitor;
+
+                impl<'de> Visitor<'de> for FieldVisitor {
+                    type Value = Field;
+
+                    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                        formatter.write_str("`chrom`, `start` or `end`")
+                    }
+
+                    fn visit_str<E>(self, value: &str) -> std::result::Result<Field, E>
+                    where
+                        E: de::Error,
+                    {
+                        match value {
+                            "chrom" => Ok(Field::Chrom),
+                            "start" => Ok(Field::Start),
+                            "end" => Ok(Field::End),
+                            _ => Err(de::Error::unknown_field(value, FIELDS)),
+                        }
+                    }
+                }
+
+                deserializer.deserialize_identifier(FieldVisitor)
+            }
+        }
+
+        struct Record3Visitor;
+
+        impl<'de> Visitor<'de> for Record3Visitor {
+            type Value = Record<3>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("struct Record<3>")
+            }
+
+            fn visit_seq<V>(self, mut seq: V) -> std::result::Result<Record<3>, V::Error>
+            where
+                V: SeqAccess<'de>,
+            {
+                let chrom: &str = seq.next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(0, &self))?;
+                let start: usize = seq.next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(1, &self))?;
+                let end: usize = seq.next_element()?
+                    .ok_or_else(|| de::Error::invalid_length(2, &self))?;
+
+                // TODO: unwrap used in order to respect V::Error
+                //          need a way to convert builder error to remove unwraps
+                //          and resume returning error if needed?
+                Ok(Record::<3>::builder()
+                    .set_reference_sequence_name(chrom)
+                    .set_start_position(Position::try_from(start).unwrap())
+                    .set_end_position(Position::try_from(end).unwrap())
+                    .build().unwrap())
```

Comment:
V::Error should extend from Rust's standard error, which means `map_err` will work with a bit of formatting. There are some functions in [`serde::de::Error`](https://docs.serde.rs/serde/de/trait.Error.html) for specific error types, or the `custom` function would work too.

For example, replace the `.unwrap()` with `.map_err(|err| de::Error::custom(err))` to turn this into the correct error type.